packages/smithy-core/src/smithy_core/endpoints.py

```diff
@@ -0,0 +1,89 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+from typing import Any
+from dataclasses import dataclass, field
+from urllib.parse import urlparse
+
+from . import URI
+from .serializers import SerializeableShape
+from .schemas import APIOperation
+from .interfaces import TypedProperties as _TypedProperties
+from .interfaces import Endpoint as _Endpoint
+from .interfaces import URI as _URI
+from .types import TypedProperties, PropertyKey
+from .exceptions import EndpointResolutionError
+
+
+STATIC_URI: PropertyKey[str | _URI] = PropertyKey(
+    key="endpoint_uri",
+    # Python currently has problems expressing parametric types that can be
+    # unions, literals, or other special types in addition to a class. So
+    # we have to ignore the type below. PEP 747 should resolve the issue.
+    # TODO: update this when PEP 747 lands
+    value_type=str | _URI,  # type: ignore
+)
+"""The property key for a statically defined URI."""
+
+
+@dataclass(kw_only=True)
+class Endpoint(_Endpoint):
+    """A resolved endpoint."""
+
+    uri: _URI
+    """The endpoint URI."""
+
+    properties: _TypedProperties = field(default_factory=TypedProperties)
+    """Properties required to interact with the endpoint.
+
+    For example, in some AWS use cases this might contain HTTP headers to add to each
+    request.
+    """
+
+
+@dataclass(kw_only=True)
+class EndpointResolverParams[I: SerializeableShape]:
```

Comment:
I do like this as a simplification for all endpoint parameters and to some extent it allows more customer extension to clients (since they can interact with any part of the input or context). However, it doesn't quite match to EP2's guidance on generating specific parameters per service.

Reply:
> it doesn't quite match to EP2's guidance on generating specific parameters per service

I very explicitly do not want to do that. It dramatically increases complexity by foisting the logic of extracting what's important into the request pipeline, which ideally knows nothing about the resolver other than that it exists. There's no way to even do it without codegen that isn't just exactly what the resolver would do when given the world, except there's extra steps because you instead have to ask for the resolver params class, give that the world, and then give that to the resolver.

All for what? A moderately easier to test interface? No thanks.

Reply:
The problem is that we've created a world where we need a broad interface. We've made the bed, we might as well admit that we're lying in it.

Reply:
Yeah thats a fair point - should we be taking the same approach for Auth resolver/parameters?